{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extending input sequences for CRISPR base editing modeling\n",
    "Caleb Lareau\n",
    "\n",
    "The way this works is that we'll import the old .fasta files, pull an extended sequence, and then export new .fasta files. The reference genome is hg38. For super long sequences, we may run into more edge cases where the base that it is being modeled runs off of the transcript and into gDNA, which doesn't necessarily represent reality on the mRNA molecule. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Run this once to get the dependencies and then delete it\n",
    "install.packages(\"data.table\")\n",
    "install.packages(\"seqinr\")\n",
    "install.packages(\"stringr\")\n",
    "install.packages(\"BiocManager\")\n",
    "BiocManager::install(\"BSgenome.Hsapiens.UCSC.hg38\")\n",
    "BiocManager::install(\"Biostrings\")\n",
    "BiocManager::install(\"GenomicRanges\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: BSgenome\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: Biostrings\n",
      "Loading required package: XVector\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "Attaching package: ‘seqinr’\n",
      "\n",
      "The following object is masked from ‘package:Biostrings’:\n",
      "\n",
      "    translate\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(Biostrings)\n",
    "library(data.table)\n",
    "library(seqinr)\n",
    "library(stringr)\n",
    "library(GenomicRanges)\n",
    "\n",
    "# Crude pass that pulls from gDNA -- does not consider any potential splicing\n",
    "getSequenceSimple <- function(gr_query, reverse_complement = FALSE){\n",
    "  \n",
    "  gs <- getSeq(BSgenome.Hsapiens.UCSC.hg38, gr_query)\n",
    "  \n",
    "  # Flip the sequence if we want to\n",
    "  if(reverse_complement){\n",
    "    gs_rc <- reverseComplement(gs)\n",
    "    gs <- gs_rc\n",
    "  }\n",
    "  \n",
    "  return(as.character(gs))\n",
    "}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Function that takes a current fasta file and then pads\n",
    "# the desired amount, resulting in a new fasta file\n",
    "# with sequence with 2*new_pad + 1\n",
    "repad_fasta <- function(path_to_fasta, new_pad = 100){\n",
    "    dt <- fread(path_to_fasta, header = FALSE)\n",
    "    headers <- gsub(\">\", \"\", dt[rep(c(TRUE,FALSE),dim(dt)[1]/2),][[1]])\n",
    "    sequences <- dt[rep(c(FALSE,FALSE),dim(dt)[1]/2),][[1]]\n",
    "\n",
    "    # Munge the headers a bit\n",
    "    split_mat <- str_split_fixed(headers, \"_\", 5)\n",
    "    genome_positions <- str_split_fixed(split_mat[,2], \"-\", 2)\n",
    "\n",
    "    granges_seqs <- makeGRangesFromDataFrame(\n",
    "        data.frame(\n",
    "        chr = genome_positions[,1],\n",
    "        start = as.numeric(genome_positions[,2]) - new_pad,\n",
    "        end = as.numeric(genome_positions[,2]) + new_pad\n",
    "        )\n",
    "    )\n",
    "    boo_forward <- split_mat[,3] == \"F\"\n",
    "    new_seqs_F <- getSequenceSimple(granges_seqs[boo_forward])\n",
    "    new_seqs_R <- getSequenceSimple(granges_seqs[!boo_forward], reverse_complement = TRUE)\n",
    "\n",
    "    # Export the sequences to a new fasta file\n",
    "    fasta_names <- c(headers[boo_forward], headers[!boo_forward])\n",
    "    new_fasta_file <- gsub(\".fasta.gz\", paste0(\"_pad\",as.character(new_pad),\".fasta\"), path_to_fasta)\n",
    "    write.fasta(as.list(c(new_seqs_F, new_seqs_R)), fasta_names, new_fasta_file, open = \"w\")\n",
    "    system(paste0(\"gzip \", new_fasta_file))\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here's a minimal function call but this is best done in a loop most likely. \n",
    "# One can use the list.files() command in R to show files in a path \n",
    "# And then loop over them. \n",
    "\n",
    "for (x in list.files(\"../data/processed/ABE/ABE-sequence/\")) {\n",
    "    repad_fasta(paste(\"../data/processed/ABE/ABE-sequence/\", x, sep = \"\"), 500)\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:r_env]",
   "language": "R",
   "name": "conda-env-r_env-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
